packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts

```diff
@@ -0,0 +1,573 @@
+   *   - if any old epochs checkpoint states are persisted, no need to do it again
+   *
+   * Note that for each epoch there could be multiple checkpoint states, usually 2, one for Previous Root Checkpoint State and one for Current Root Checkpoint State.
+   * We normally only persist 1 checkpoint state per epoch, the one that could potentially be justified/finalized later based on the view of the block.
+   * Other checkpoint states are pruned from memory.
+   *
+   * This design also covers the reorg scenario. Given block c in the same epoch n where c.slot > b.slot, c is not descendant of b, and c is built on top of c0
+   * instead of b0 (epoch (n - 1))
+   *
+   *       epoch:          (n-2)       (n-1)         n         (n+1)
+   *             |-----------|-----------|-----------|-----------|
+   *                        ^            ^       ^   ^    ^   ^
+   *                        |            |       |   |    |   |
+   * block chain:           b2---------->b1----->c0->b0-->b   |
+   *                                             ║            |
+   *                                             ╚═══════════>c (reorg)
+   *
+   * After processing block c, if maxEpochsInMemory is:
+   * - 0 then we'll persist {root: c0, epoch: n} checkpoint state to disk. Note that regen should populate {root: c0, epoch: n} checkpoint state before.
+   *
+   *                           epoch:      (n-1)                                                           n                                                           (n+1)
+   *                                         |-------------------------------------------------------------|-------------------------------------------------------------|
+   *                                         ^                               ^                             ^                             ^
+   *   _______                               |                               |                             |                             |
+   *  |       |                              |                               |                             |                             |
+   *  |  db   |====== reload ======> {root: b1, epoch: n-1} cp state ======> c0 block state ======> {root: c0, epoch: n} cp state =====> c block state
+   *  |_______|
+   *
+   *
+   *
+   * - 1 then we'll persist {root: b1, epoch n-1} checkpoint state to disk. Note that at epoch n there is both {root: b0, epoch: n} and {root: c0, epoch: n} checkpoint states in memory
+   * - 2 then we'll persist {root: b2, epoch n-2} checkpoint state to disk, there are also 2 checkpoint states in memory at epoch n, same to the above (maxEpochsInMemory=1)
+   *
+   * As of Nov 2023, it takes 1.3s to 1.5s to persist a state on holesky on fast server. TODO:
+   * - improve state serialization time
+   * - or research how to only store diff against the finalized state
+   */
+  async processState(blockRootHex: RootHex, state: CachedBeaconStateAllForks): Promise<number> {
```

Comment:
please add vertical whitespace in this function, it's quite dense and hard to follow?